### ISSUE TITLE
128 bits missing from return

### ISSUE THREAD

**Issue by adam-street** [mscdex/ssh2#733]
I am connecting to a old unix server that is returning cp437 when i load a program screen its returning 
0x44 (68) for a 0xC4(196) character. For some reason its chopping 128 bits of the return.

`
.on('data', function(buffer) {

      console.log(buffer);

      var hexString = buffer.toString('hex');
      var hexArray = hexString.match(/.{2}/g);
      hexArray.forEach(hex => {
        fs.appendFile('putty-data-string.txt', `${hex} `, function (err) {
          if (err) throw err;
        });
      });

      fs.appendFile('putty-data-string.txt', `\n`, function (err) {
        if (err) throw err;
      });
`

Is there a setting that i can change to get access to the missing bits?

**Comment by mscdex** [comment]
1. Are you calling `.setEncoding()` on the stream?
2. Don't expect to receive every byte of data you're looking for in a single `'data'` event. Did you verify that your `'data'` event handler is only called once (is it possible the missing data is made available in the next `'data'` event)? Did you also verify `buffer.length >= 16` (128 bits)?
3. Be careful with asynchronous operations to the same file, do not assume that they will be processed in order. Is there a reason why you are issuing separate append requests for every byte of data?

**Comment by adam-street** [comment]
First off,
 Thanks for the quick reply!

1. refer to new code below
2. 'data' is only called once, the buffer length is well over 128 is most cases please refer to output below
3. the file stuff is just for debugging, just wanted to see a print out of hex in a file with space between, ive corrected this in the new code (thanks for the tip)

New code
`var Client = require('ssh2').Client;

var fs = require('fs');
 
var conn = new Client();
conn.on('ready', function() {
  console.log('Client :: ready');
  conn.shell(function(err, stream) {
    if (err) throw err;
    stream.setEncoding('hex');
    stream.on('close', function() {
      console.log('Stream :: close');
      conn.end();
    }).on('data', function(buffer) {

      var hexStr = buffer.toString('hex');
      console.log(hexStr);

      // This could be wrong, but 1 hex char = 4 bits and it is returning 2 char hex per buffer item
      if (buffer.length * 8 >= 256) {
        fs.appendFile('putty-data-string.txt', `${hexStr} \n`, function (err) {
          if (err) throw err;
        });
      } 
      else {
        console.log('buffer less than 126 bits')
        fs.appendFile('putty-data-string.txt', `Buffer less than 126 bits: ${hexStr} \n`, function (err) {
          if (err) throw err;
        });
      }

    }).stderr.on('data', function(data) {~
      console.log('STDERR: ' + data);
    });

    // Run program
    stream.write('ansi-putty\n. dev\ncd src/cs\nrun order\n');

  });
}).connect({
  host: '*******',
  port: 22,
  username: '******',
  password: '******',
  tryKeyboard: true
});`

Example output
`1b5b421b5b421b5b31326d1b5b316d1b5b34306d1b5b33376d4444444444444444444444444444444444444444444444444444444444444444444444444444444444444444444444444444444444444444444444444444441b5b31306d1b5b34316d1b5b33376d204f72646572201b5b31326d1b5b34306d1b5b33376d44444444444444444444444444444444444444444444444444444444444444444444444444444444444444444444441b5b31306d1b5b34336d1b5b33306d51756572793a2043757272656e741b5b31326d1b5b34306d1b5b33376d440d0a1b5b31306d1b5b34306d1b5b33366d4469762f50726f2e2e3a201b5b33336d5f5f1b5b3336 
6d2d1b5b33336d5f5f5f5f5f5f1b5b33373b34306d1b5b31303b306d201b5b316d1b5b34306d1b5b33336d20202020202020202020201b5b33373b34306d1b5b31303b306d201b5b316d1b5b34306d1b5b33366d52676e3a1b5b33373b34306d1b5b31303b306d201b5b316d1b5b34306d1b5b33336d5f5f5f1b5b33373b34306d1b5b31303b306d201b5b316d1b5b34306d1b5b33366d4f72646572233a1b5b33373b34306d1b5b31303b306d201b5b316d1b5b34306d1b5b33336d5f5f5f5f5f5f5f5f5f1b5b33373b34306d1b5b31303b306d201b5b316d1b5b34306d1b5b33366d42696c6c2d546f2e3a1b5b33373b34306d1b5b31303b306d201b5b316d 
1b5b34306d1b5b33336d5f5f5f5f5f5f1b5b33373b34306d1b5b31303b306d201b5b316d1b5b34306d1b5b33336d202020202020202020202020202020202020202020202020201b5b33373b34306d1b5b31303b306d201b5b316d1b5b34306d1b5b33366d5174793a1b5b33373b34306d1b5b31303b306d201b5b316d1b5b34306d1b5b33336d5f5f5f1b5b33373b34306d1b5b31303b306d0920201b5b316d1b5b34306d1b5b33366d496e766f6963651b5b33373b34306d1b5b31303b306d2020202020201b5b316d1b5b34306d1b5b33366d543130301b5b33373b34306d1b5b31303b306d0d1b5b421b5b316d1b5b34306d1b5b33366d4f726465725479 
7`

The 0x44 here is the issue i know the server is returning a 196 char code (0xc4) but its losing some  bits somewhere along the way.

**Comment by mscdex** [comment]
1. Using `shell()` for automating commands is generally not the best solution. Use `exec()` if you can (if your commands require a tty, you can enable this with `exec()` as well) as it will not include output from the shell and other output that might get in your way.

2. You're setting the stream's encoding to `'hex'` already, so calling `buffer.toString('hex')` is redundant and only works because a JavaScript string's `.toString()` method ignores any arguments passed to it.

3. I suggest simplifying things and getting rid of all of the hex-related logic. Try something like this:

```js
conn.shell((err, stream) => {
  if (err) throw err;
  stream.on('data', (chunk) => {
    console.log(`chunk: ${chunk.length} bytes / ${chunk.length * 8} bits`);
    console.log(`... as hex string: ${chunk.toString('hex')}`);
    console.log('... as utf8 string: ${require('util').inspect(chunk.toString('utf8'))}`);
  });
  stream.write('ansi-putty\n. dev\ncd src/cs\nrun order\n');
});
```

**Comment by adam-street** [comment]
1. End goal here is to display a term in a browser. The only automation i will be doing is starting up a specific program on the server. The program is interactive and will require display / user input

2. LOL facepalm, thanks for the info

3. tried the simplified logic and still running into the same issue.

Code:
`        var fs = require('fs');
        var Client = require('ssh2').Client;
        var conn = new Client();

        conn.on('ready', function() {
          conn.shell((err, stream) => {
            if (err) throw err;
            stream.on('data', (chunk) => {

              var rtn = `chunk: ${chunk.length} bytes / ${chunk.length * 8} bits\n
                        ... as hex string: ${chunk.toString('hex')}\n
                        ... as utf8 string: ${require('util').inspect(chunk.toString('utf8'))}\n`;

              console.log(rtn);

              fs.appendFile('putty-data-string.txt', rtn, function (err) {
                if (err) throw err;
              });
            });
            stream.write('ansi-putty\n. dev\ncd src/cs\nrun order\n');
          });
        }).connect({
          host: '******',
          port: 22,
          username: '******',
          password: '******',
          tryKeyboard: true
        });
` 

Output sample: 
`
        chunk: 256 bytes / 2048 bits

                        ... as hex string: 1b5b31306d1b5b34316d1b5b33376d204f726465722053746f70732f496e666f726d6174696f6e201b5b31326d1b5b34306d1b5b33376d44444444444444444444444444444444444444444444444444444444444444444444444444444444444444444444444444444444440d0a1b5b33373b34306d1b5b31303b306d201b5b316d1b5b34306d1b5b33366d547261696c6572204f726967696e3a1b5b33373b34306d1b5b31303b306d201b5b316d1b5b34306d1b5b33336d5f5f5f5f5f5f1b5b33373b34306d1b5b31303b306d090920201b5b316d1b5b34306d1b5b33366d536869707065723a1b5b33373b34306d1b5b31303b306d201b5b316d1b5b3430

                        ... as utf8 string: '\u001b[10m\u001b[41m\u001b[37m Order Stops/Information \u001b[12m\u001b[40m\u001b[37mDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDD\r\n\u001b[37;40m\u001b[10;0m \u001b[1m\u001b[40m\u001b[36mTrailer Origin:\u001b[37;40m\u001b[10;0m \u001b[1m\u001b[40m\u001b[33m______\u001b[37;40m\u001b[10;0m\t\t  \u001b[1m\u001b[40m\u001b[36mShipper:\u001b[37;40m\u001b[10;0m \u001b[1m\u001b[40'

`

still have the same issue where its returning 0x44 not 0xC4

**Comment by mscdex** [comment]
Does it show the correct output when you ssh in via the OpenSSH client? If so, what does `echo $TERM` show for you? Perhaps you will need to explicitly set which terminal to use for the tty/pty. By default `'vt100'` is used, so that could explain why you're not seeing fancier output.

FWIW if you're doing this as part of a browser-based session, you might consider trying [this example](https://stackoverflow.com/a/38690242/2050455) and tweaking the backend connection part to your needs (e.g. adding your forced `stream.write(...)` inside the `shell()` callback).

**Comment by stfortner93** [comment]
Hey, I am working with Adam, and we figured out the issue. 

It is that the linux server sends us the escape sequence '\u001b[12m', that tells us to use an alternate font for the next non-escape sequence character, and then '\u001b[10m' tells us to switch back to normal font. \u001b is just the escape character.  

For us, what is the best way for us to always gather that data, and then display it correctly?

**Comment by mscdex** [comment]
You have access to the raw binary data as long as you don't use `setEncoding()`. From there it's just a matter of searching for your escape sequences. However, you might consider explicitly setting the `term` option and using a project like xterm.js if you are building something for display in the browser, as it should take care of interpreting all of the escape sequences for you.

**Comment by adam-street** [comment]
Solved! Thanks for the help was not an issue on the ssh2 side at all! Sorry! We are still learning when it comes to how terminals work under the hood. We had to create a custom term cap file on the server so we would receive the correct command sequences to send to xterm.js to render color and line drawing character properly. Your software is working flawlessly, thanks again for creating such a useful lib, and for your quick, helpful, and friendly responses.
